src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs

```diff
@@ -1251,4 +1251,1192 @@ static class C
+            }
+            """;
+        // PROTOTYPE: Should not be an error when value checks are relaxed. Verify execution and IL.
+        CreateCompilation(source).VerifyDiagnostics(
+            // (5,16): error CS8329: Cannot use variable 'p' as a ref or out value because it is a readonly variable
+            //         M2(ref p);
+            Diagnostic(ErrorCode.ERR_RefReadonlyNotField, "p").WithArguments("variable", "p").WithLocation(5, 16));
+    }
+
+    [Fact]
+    public void PassingParameters_In_RefReadonly_InArgument()
+    {
+        var source = """
+            class C
+            {
+                static void M1(in int p)
+                {
+                    M2(in p);
+                }
+                static void M2(ref readonly int p) => System.Console.Write(p);
+                static void Main() => M1(5);
+            }
+            """;
+        var verifier = CompileAndVerify(source, expectedOutput: "5");
+        verifier.VerifyDiagnostics();
+        verifier.VerifyIL("C.M1", """
+            {
+              // Code size        7 (0x7)
+              .maxstack  1
+              IL_0000:  ldarg.0
+              IL_0001:  call       "void C.M2(ref readonly int)"
+              IL_0006:  ret
+            }
+            """);
+    }
+
+    [Fact]
+    public void PassingParameters_RefReadonly_In_PlainArgument()
+    {
+        var source = """
+            class C
+            {
+                static void M1(ref readonly int p)
+                {
+                    M2(p);
+                }
+                static void M2(in int p) => System.Console.Write(p);
+                static void Main()
+                {
+                    int x = 5;
+                    M1(ref x);
+                }
+            }
+            """;
+        var verifier = CompileAndVerify(source, expectedOutput: "5");
+        verifier.VerifyDiagnostics();
+        verifier.VerifyIL("C.M1", """
+            {
+              // Code size        7 (0x7)
+              .maxstack  1
+              IL_0000:  ldarg.0
+              IL_0001:  call       "void C.M2(in int)"
+              IL_0006:  ret
+            }
+            """);
+    }
+
+    [Fact]
+    public void PassingParameters_RefReadonly_In_RefArgument()
+    {
+        var source = """
+            class C
+            {
+                static void M1(ref readonly int p)
+                {
+                    M2(ref p);
+                }
+                static void M2(in int p) => System.Console.Write(p);
+                static void Main()
+                {
+                    int x = 5;
+                    M1(ref x);
+                }
+            }
+            """;
+        var verifier = CompileAndVerify(source, expectedOutput: "5");
+        verifier.VerifyDiagnostics(
+            // (5,16): warning CS9502: The 'ref' modifier for argument 1 corresponding to 'in' parameter is equivalent to 'in'. Consider using 'in' instead.
```

Comment:
>// (5,16): warning CS9502: The 'ref' modifier for argument 1 corresponding to 'in' parameter is equivalent to 'in'. Consider using 'in' instead. [](http://example.com/codeflow?start=12&length=144)

I think inside M1 we should treat p as a readonly reference (similar to how `in` parameter would be treated). Therefore, we should disallow usage of `ref` modifier with it and report: "Cannot use variable 'p' as a ref or out value because it is a readonly variable"

Reply:
We do treat it as a readonly reference, the `ref` argument modifier doesn't change that (same as above).